### PR TITLE
[kube-prometheus-stack] prometheus-operator0.59.1/Prometheus2.38.0/all dependencies

### PR DESCRIPTION
**Pull request by Allex1**
#### Which issue this PR fixes

*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*

- Bumps the kube-state-metrics dependency version to latest

#### Checklist

- [x] [DCO](https://github.com/prometheus-community/helm-charts/blob/main/CONTRIBUTING.md#sign-off-your-work) signed
- [x] Chart Version bumped
- [x] Title of the PR starts with chart name (e.g. `[prometheus-couchdb-exporter]`)

### CODE REVIEW

**Review by monotek — CHANGES_REQUESTED** [review]
this should be a breaking changhe as the default lablebs are changed in the subchart.
please add update instructions to readme.

File: charts/kube-prometheus-stack/Chart.lock [hunk 1]
```diff
@@ -10,3 +10,4 @@ dependencies:
   version: 6.37.3
 digest: sha256:1db99e2752cc271323cf886c3ea14105f3aec0883ec67e944533f6246870d15c
 generated: "2022-09-11T15:48:48.5400618+03:00"
+
```
Comment (mrueg) [inline charts/kube-prometheus-stack/Chart.lock:13]: Looks like this blank line got added by accident.

Comment (Allex1) [reply charts/kube-prometheus-stack/Chart.lock]: fixed

**Comment by Allex1** [comment]
> this should be a breaking changhe as the default labels are changed in the subchart.
> please add update instructions to readme.

@monotek  I'm assuming you're referring to https://github.com/prometheus-community/helm-charts/pull/927

**Comment by Allex1** [comment]
@mrueg @monotek because 40.0.0 is such a round number I've updated prom-op and prom as well.

**Review by monotek — CHANGES_REQUESTED** [review]
the crds, dashboards and rules are not upgraded (see scripts in hack directory).

the readme should have upgrade instructions for the labels, not just a note that it will break.

edit: sorry, i had node exporter in mind: https://github.com/prometheus-community/helm-charts/pull/2212

please update it too. upgrade instructions can be used from its readme: https://github.com/prometheus-community/helm-charts/tree/main/charts/prometheus-node-exporter#3x-to-4x

please test the update at least once .

File: charts/kube-prometheus-stack/Chart.lock [hunk 2]
```diff
@@ -1,12 +1,12 @@
 dependencies:
 - name: kube-state-metrics
   repository: https://prometheus-community.github.io/helm-charts
-  version: 4.15.0
+  version: 4.18.0
 - name: prometheus-node-exporter
   repository: https://prometheus-community.github.io/helm-charts
   version: 3.3.1
```
Comment (mrueg) [inline charts/kube-prometheus-stack/Chart.lock:7]: do you want to update node exporter as well https://github.com/prometheus-community/helm-charts/commits/main/charts/prometheus-node-exporter

**Review by mrueg — COMMENTED** [review]
values.yaml needs the prometheus-operator and config-reloader tags updated to v0.59.1 as well

File: charts/kube-prometheus-stack/Chart.lock (hunk 2 above)
Comment (Allex1) [reply charts/kube-prometheus-stack/Chart.lock]: done

**Comment by mrueg** [comment]
Thanos can be updated to https://github.com/thanos-io/thanos/releases/tag/v0.28.0 as well

File: charts/kube-prometheus-stack/README.md [hunk 3]
```diff
@@ -80,6 +80,34 @@ _See [helm upgrade](https://helm.sh/docs/helm/helm_upgrade/) for command documen
...
+
+This version updates kube-state-metrics to 4.18.0 and prometheus-node-exporter to 4.1.0 which adds Kubernetes recommended labels on all resources.
+
+If you use your own custom [ServiceMonitor](https://github.com/prometheus-operator/prometheus-operator/blob/main/Documentation/api.md#servicemonitor) or [PodMonitor](https://github.com/prometheus-operator/prometheus-operator/blob/main/Documentation/api.md#podmonitor), please ensure to upgrade their `selector` fields accordingly to the new labels.
+
+Starting from prometheus-node-exporter version 4.0.0, the `node exporter` chart is using the [Kubernetes recommended labels](https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/). Therefore you have to delete the daemonset before you upgrade.
+
+```console
+kubectl delete daemonset -l app=prometheus-node-exporter
+helm upgrade -i prometheus-node-exporter prometheus-community/prometheus-node-exporter
```
Comment (monotek) [inline charts/kube-prometheus-stack/README.md:93]: you can't copy this 1 to 1. 
please alos test it!
we don't want to install prometheus-node-exporter standalone, which would happen with this command.

File: charts/kube-prometheus-stack/README.md [hunk 4]
```diff
@@ -80,6 +80,34 @@ _See [helm upgrade](https://helm.sh/docs/helm/helm_upgrade/) for command documen
 
 A major chart version change (like v1.2.3 -> v2.0.0) indicates that there is an incompatible breaking change needing manual actions.
 
+### From 39.x to 40.x
+
+This version updates kube-state-metrics to 4.18.0 and prometheus-node-exporter to 4.1.0 which adds Kubernetes recommended labels on all resources.
+
+If you use your own custom [ServiceMonitor](https://github.com/prometheus-operator/prometheus-operator/blob/main/Documentation/api.md#servicemonitor) or [PodMonitor](https://github.com/prometheus-operator/prometheus-operator/blob/main/Documentation/api.md#podmonitor), please ensure to upgrade their `selector` fields accordingly to the new labels.
+
+Starting from prometheus-node-exporter version 4.0.0, the `node exporter` chart is using the [Kubernetes recommended labels](https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/). Therefore you have to delete the daemonset before you upgrade.
```
Comment (monotek) [inline charts/kube-prometheus-stack/README.md:89]: please give exact instructions how to configure correct labels.

maybe we even need some other pr for nodeexporter first, adding something like: https://github.com/prometheus-community/helm-charts/blob/main/charts/kube-state-metrics/values.yaml#L39-L40

**Review by monotek — CHANGES_REQUESTED** [review]

**Comment by Allex1** [comment]
@monotek @mrueg seems this needs some more work/testing. 
Before we finish this can you guys pls have a look at https://github.com/prometheus-community/helm-charts/pull/2430 and https://github.com/prometheus-community/helm-charts/pull/2429
To me it would make sense to merge them first.

File: charts/kube-prometheus-stack/README.md (hunk 3 above)
Comment (Allex1) [reply charts/kube-prometheus-stack/README.md:93]: I'll spin up a kind cluster and give it a go as it's not really clear what are the upgrade issue.

**Comment by monotek** [comment]
> @monotek @mrueg seems this needs some more work/testing. Before we finish this can you guys pls have a look at #2430 and #2429 To me it would make sense to merge them first.

the mentioned prs are added.

i've created https://github.com/prometheus-community/helm-charts/pull/2453 to be able to use "releaseLabel: true" in node-exporter to get the needed labels added.

**Review by monotek — CHANGES_REQUESTED** [review]

**Review by monotek — COMMENTED** [review]
i've added some changes to the readme, updated some more image tags, updated node exporter to 4.2.0 and added the release label to the values. please check if update works as described.

**Comment by Allex1** [comment]
@monotek I hope to find some time and test the upgrade in the following days.

**Review by monotek — APPROVED** [review]

**Comment by monotek** [comment]
update in a kind cluster from 39.0.0 worked for me.

**Comment by Allex1** [comment]
> update in a kind cluster from 39.0.0 worked for me. had to use --force-conflicts on some crds though.
> 
> @mrueg do you see any issues?

Tested as well in kind and confirmed it's working fine. Should we add the `--force-conflicts` flag in the upgrade docs ?

**Comment by monotek** [comment]
> > update in a kind cluster from 39.0.0 worked for me. had to use --force-conflicts on some crds though.
> > @mrueg do you see any issues?
> 
> Tested as well in kind and confirmed it's working fine. Should we add the `--force-conflicts` flag in the upgrade docs ?

Not sure. Maybe it's unwanted to overwrite directly.

**Comment by totoroot** [comment]
> i've added some changes to the readme, updated some more image tags, removed some unneeded comments, updated node exporter to 4.2.0 and added the release label to the values. please check if update works as described.

@monotek With your changes added to this PR you basically just removed comments which were added with PR #2411, because you deemed them to be not useful although @moroviintaas and @Xtigyro disagreed.

The chart's values.yaml has over 3000 LOC and I agree that this might not be ideal, but removing comments other consider helpful isn't the way to go IMO. I'd personally wish for changes like that to be done more transparently.

**Pull request merged**